src/Common/ThreadProfileEvents.cpp

```diff
@@ -104,6 +118,401 @@ void TasksStatsCounters::incrementProfileEvents(const ::taskstats & prev, const
+            fd_count_afterwards, threshold, maximum_open_descriptors,
+            FILE_DESCRIPTORS_THRESHOLD);
+        return false;
+    }
+
+    // open descriptors for new events
+    for (size_t i : events_to_open)
+    {
+        const PerfEventInfo & event_info = raw_events_info[i];
+        int & fd = thread_events_descriptors_holder.descriptors[i];
+        // disable by default to add as little extra time as possible
+        fd = openPerfEventDisabled(perf_event_paranoid, has_cap_sys_admin, event_info.event_type, event_info.event_config);
+
+        if (fd == -1)
+        {
+            LOG_WARNING(getLogger(), "Perf event is unsupported: {}"
+                " (event_type={}, event_config={})",
+                event_info.settings_name, event_info.event_type,
+                event_info.event_config);
+        }
+    }
+
+    return true;
+}
+
+// Parse comma-separated list of event names. Empty or 'all' means all available
+// events.
+// TODO add validation to setting
+std::vector<size_t> PerfEventsCounters::eventIndicesFromString(const std::string & events_list)
+{
+    std::unordered_set<std::string> requested_events;
+    std::istringstream iss(events_list);
+    std::string event_name;
+    while (std::getline(iss, event_name, ','))
+    {
+        requested_events.insert(event_name);
+    }
+
+    std::vector<size_t> result;
+    result.reserve(PerfEventsCounters::NUMBER_OF_RAW_EVENTS);
+    if (requested_events.size() == 0
+        || requested_events.count("all") > 0)
+    {
+        for (size_t i = 0; i < PerfEventsCounters::NUMBER_OF_RAW_EVENTS; ++i)
+        {
+            result.push_back(i);
+        }
+        return result;
+    }
+
+    for (size_t i = 0; i < PerfEventsCounters::NUMBER_OF_RAW_EVENTS; ++i)
+    {
+        if (requested_events.count(
+            PerfEventsCounters::raw_events_info[i].settings_name) > 0)
+        {
+            result.push_back(i);
+        }
+    }
+
+    return result;
+}
+
+void PerfEventsCounters::initializeProfileEvents(PerfEventsCounters & counters, const std::string & events_list)
+{
+    if (current_thread_counters_id.has_value())
+    {
+        if (current_thread_counters_id != counters.id)
+            LOG_WARNING(getLogger(), "Only one instance of `PerfEventsCounters` can be used on the thread");
+        return;
+    }
+
+    if (!processThreadLocalChanges(events_list))
+        return;
+
+    for (int fd : thread_events_descriptors_holder.descriptors)
+    {
+        if (fd == -1)
+            continue;
+
+        resetPerfEvent(fd, getLogger);
+        enablePerfEvent(fd, getLogger);
+    }
+
+    current_thread_counters_id = counters.id;
+}
+
+void PerfEventsCounters::finalizeProfileEvents(PerfEventsCounters & counters, ProfileEvents::Counters & profile_events)
+{
+    if (current_thread_counters_id != counters.id)
+        return;
+
+    const auto old_values = counters.raw_event_values;
```

Comment:
Uff, previous review was accidentally deleted. Assigning `old_values` in this case does not copy values but only the pointer to them, thus these values will be rewritten with the new ones on `read(...)`

Reply:
And another one... There will be no warning if another instance of `PerfEventCounters` will be used for the previous query, which can lead to incorrect results (though, there is no problem for now)

Reply:
Thanks, fixed.